### DISCUSSION TITLE
Shipping the `png` crate in Chromium

### DISCUSSION THREAD

**Discussion by anforowicz** [image-rs/image-png#562] [General]
Hello!

I really appreciate all the work that has been put into the `png` crate (and `fdeflate` crate as well as the Rust toolchain and other Rust crates).  As a way of saying "thank you", let me share the status (and excitement) of what is going on with shipping the `png` crate in Chromium.

History of edits of this post:
* 2024-12-30: initial post
* 2025-02-21: updated to reflect newly started Canary+Dev field trials
* 2025-03-13: updated to reflect newly started Beta field trials
* 2025-03-27: updated to reflect postponing the launch to M136
* 2025-04-16: updated to reflect 1% Stable field trials that will start in 1-2 weeks

We started Chrome Canary experiments in the 2nd half of December 2024, Chrome Dev experiments in the 2nd half of February 2025, and Chrome Beta experiments in the middle of March 2023.  We plan to enable Rust PNG for 1% of users of the Stable release channel when M136 ships to that release channel at the end of April.

Below are some additional details in the form of questions and answers.  Let me know if you have any other questions.

------------

# Q: What Chrome versions participate in the experiments?

- Chrome [133.0.6891.0](https://chromiumdash.appspot.com/commit/d24817349336bb6086d78afdf418fed389a6a1d1) is the first version with Rust PNG built into the binary.  Rust PNG is disabled by default in this version, but can be force-enabled using chrome://flags/#rusty-png.
- Chrome 133.0.6892.0 was the minimum version in the current Canary 50%-50% experiments in December 2024.
- Chrome 136.0.7093.0 is the minimum version for participating in the current Canary+Dev+Beta 50%-50% experiments and for the planned Stable 1% experiment.

For a tentative Chromium release schedule please see https://chromiumdash.appspot.com/schedule.

# Q: How can I tell which experiment group I am in?

To find out which experiment group you are in, go to chrome://version and search for “977e3a5f” in the list of “Active Variations”:

- Control groups:
    - 977e3a5f-89078435 maps to RustyPng-Control_20241212 (Canary experiments in Dec 2024 - Jan 2025)
    - 977e3a5f-05a8121a maps to RustyPng-Control_20250220 (Canary+Dev experiments in Feb-... 2025)
    - 977e3a5f-dd8cc695 maps to RustyPng-Control_20250312 (M135+ Canary+Dev+Beta experiments in March-... 2025)
    - 977e3a5f-b8767ac5 maps to RustyPng-Control_20250327 (M136+ Canary+Dev+Beta experiments in March-... 2025)
    - 977e3a5f-602e4a00 maps to RustyPng-Control_20250415 (M136+ Canary+Dev+Beta+Stable experiments in April-... 2025)
- Enabled groups:
    - 977e3a5f-2a45d995 maps to RustyPng-Enabled_20241212
    - 977e3a5f-c1958267 maps to RustyPng-Enabled_20250220
    - 977e3a5f-af2a6a20 maps to RustyPng-Enabled_20250312
    - 977e3a5f-8cfe3e08 maps to RustyPng-Enabled_20250327
    - 977e3a5f-d800d39f maps to RustyPng-Enabled_20250415

Note that getting assigned to one of the experiment groups may depend on getting a new field trials “seed” and therefore may require rebooting a browser.  Also note that your local [ChromeVariations enterprise policy](https://chromeenterprise.google/policies/#ChromeVariations) may limit which field trials you may participate in.

# Q: Can I switch to a different experiment group?

You can't pick which experiment group you will be assigned to.  Experiment group assignment happens randomly, and is “sticky” / “permanent” for the given Chrome client.

If you want to force Rust `png` on or off (e.g. to triage if the field trial may be responsible for a bug), then navigate to chrome://flags/#rusty-png and force the flag to be enabled or disabled.  Note that this will exclude you from the A/B comparisons (i.e. you will be assigned to “ForcedOn_…” or “ForcedOff_…” group, rather than to the control or the experiment group).

# Q: What versions of the libraries are covered?

- Control group: `libpng`/`zlib`-based decoders and encoders that normally ship in Chromium
    - [`blink::PNGImageDecoder`](https://source.chromium.org/chromium/chromium/src/+/main:third_party/blink/renderer/platform/image-decoders/png/png_image_decoder.h;l=39;drc=1fa4135d927c4b9eaec3d0e14cb30199c22dfeaf)
    - [`SkPngDecoder`](https://source.chromium.org/chromium/chromium/src/+/main:third_party/skia/include/codec/SkPngDecoder.h;drc=14dceac9d2760735261351ed9b2a3834fed7f8a3) (used mostly through [`//ui/gfx/codec/png_codec.h`](https://source.chromium.org/chromium/chromium/src/+/main:ui/gfx/codec/png_codec.h;drc=92272b8fbd80b95c7539e7853b13fbe726b035f9) API)
    - [`SkPngEncoder`](https://source.chromium.org/chromium/chromium/src/+/main:third_party/skia/include/encode/SkPngEncoder.h;drc=0d94e966268bbc200ea57e074f33afca6321e483) (used mostly through [`//ui/gfx/codec/png_codec.h`](https://source.chromium.org/chromium/chromium/src/+/main:ui/gfx/codec/png_codec.h;drc=92272b8fbd80b95c7539e7853b13fbe726b035f9) API)
- Experiment group: Chromium/Skia’s decoders and encoders based on [Rust `png` crate](https://docs.rs/png)
    - [`SkPngRustDecoder`](https://source.chromium.org/chromium/chromium/src/+/main:third_party/skia/experimental/rust_png/decoder/SkPngRustDecoder.h;drc=14f8f6d984ff23bed1f9ae7acf98be683eccaeac) and [`SkPngRustEncoder`](https://source.chromium.org/chromium/chromium/src/+/main:third_party/skia/experimental/rust_png/encoder/SkPngRustEncoder.h;drc=e11d707452240d4d69910aa6f4e815b4f0420eea)
    - 136.0.7093.0+: `png` version: 0.18.0-rc;  `fdeflate` version: 0.3.7;  see also Chromium’s [`Cargo.lock`](https://source.chromium.org/chromium/chromium/src/+/main:third_party/rust/chromium_crates_io/Cargo.lock;l=930-940;drc=1d0f568d2922c30ba3bfaf17aaa48056e2d08e27).

# Q: Can you share performance results from the field trials?

Yes, I plan to share aggregate performance numbers.  OTOH, I want to wait until we have results from the Stable channel, because:
- The bigger user population of the Stable channel helps to get more statistically significant results (e.g. tighter confidence intervals).
- The numbers need to go through an internal review before being approved for sharing externally.  (Mostly to double-check that I won’t accidentally share any secret or competitive data like relative user populations on different platforms, and will only share relative rather than absolute counts or times.)

Preliminary M135/Canary+Dev data shows decoding runtime improvement on Desktop, and a regression on Mobile/ARM.  Encoding performance results are similarly mixed, but in the other direction.  These 2 aspects should get a bit better in M136 thanks to http://review.skia.org/962878, https://crrev.com/c/6333255, and https://crrev.com/c/6354419.

Disclaimer: Let me also note various factors that may impact the performance results (and therefore explain if/why Chromium numbers may not match other measurements):
- Chromium uses a [patched](https://source.chromium.org/chromium/chromium/src/+/main:third_party/zlib/patches/;drc=67c52873de10c06808c0de985a1e3cb8c3456d7a) version of `zlib` that uses SIMD for better performance (and therefore other results comparing across `libpng` and Rust PNG may not necessarily carry over to this A/B experiment in Chromium)
- Palette expansion optimizations in [`libpng`](https://source.chromium.org/chromium/chromium/src/+/main:third_party/libpng/arm/palette_neon_intrinsics.c;drc=ebf6d473ea1ee1963abd3afa4e18719b66be0719) or [`png` crate](https://github.com/image-rs/image-png/pull/462) are not used by `SkPngDecoder` nor `SkPngRustDecoder`, because (unlike `blink::PNGImageDecoder`) [Skia applies](https://source.chromium.org/chromium/chromium/src/+/main:third_party/skia/src/codec/SkPngCodecBase.cpp;l=343;drc=905e123aa1f0a42d1f2c26a00a9bf06870a48670) alpha premultiplication and color space transformations to <=256 pixels of the palette, rather than to N pixels of the image.
- Chromium ships a single binary per platform, which means that the compiler can only target the baseline instruction set.  For example, auto-vectorization in Chromium binaries cannot use AVX-512 SIMD.
- Performance measurements are highly sensitive to the data corpus.

**Comment by yisibl** [comment]
I'm reporting a [Chrome bug #391914245](https://issues.chromium.org/issues/391914245) which will cause our site to not be able to decode [this apng](https://img.alicdn.com/imgextra/i4/O1CN01zGVFUQ1DULVY3QaNn_!!6000000000219-54-tps-1000-563.apng).

**Reply by fintelia** [reply]
Running the image through pngcheck reports "tRNS not allowed in RGB+alpha image". So the image is corrupt, but other decoders suppress the error. We've got #525 tracking changes to match that behavior, but it should probably be prioritized. Thanks for flagging this!

**Reply by anforowicz** [reply]
@fintelia, would it be possible to release a new version of the `png` crate to  https://crates.io/crates/png?  This would help me absorb the most recent fixes (including https://github.com/image-rs/image-png/pull/569) into Chromium.  I'd be happy with either a minor version update (e.g. 0.17.17) or with a new major version (e.g. 0.18.0).  Thank you in advance for your help!

(We are restarting field trials of using Rust `png` in Chrome's Canary and Dev channels - I've edited the top-level post to reflect this.  It would be great if a new `png` version can be published within a week or so to avoid impacting Dev users with known issues.  Please shout if this schedule won't work - in that case I'll try to crate a local patch that Chromium can carry in the meantime.  FWIW https://chromiumdash.appspot.com/schedule says that Chrome M135 will branch on Monday, March 3rd.)

**Reply by fintelia** [reply]
We've already taken in a bunch of API breaking changes, so it'll have to be an 0.18 release. I'll see about figuring out what's left for that and try to get a release prepared

**Reply by fintelia** [reply]
@anforowicz just opened https://github.com/image-rs/image-png/pull/577

**Reply by anforowicz** [reply]
Thanks!  I'll track Chromium updates in https://crbug.com/400455848

**Comment by anforowicz** [comment]
I just wanted to drop a quick note that we've expanded the field trials to 50% of M135/Beta.  I've edited the top-level post accordingly.

**Comment by anforowicz** [comment]
Rust PNG got approval to launch a 1% Stable experiment in M136.  The Stable channel usually ramps up more slowly, but if everything goes smoothly we should have some data in the 2nd half of May.  I've edited the top-level post to also cover Stable trials.

**Comment by anforowicz** [comment]
🎶 Now these points of data
make a beautiful line.
And we're out of beta.
We're releasing on time. 🎶

I am happy to share that the M136 data supports shipping Rust PNG to all Chrome users.  Thank you very much to all the open source contributors whose work made it possible to ship Rust PNG with minimal performance impact and with no functional bugs so far.  Below I am sharing some high-level performance data from the rollout of Rust PNG in Chrome M136 - I hope that this data will be helpful in guiding future performance work.

Note that shipping to all Chrome users will be done gradually - initially through field trial configuration and in the long-term by enabling Rust PNG by default and then removing `libpng` from Chrome.  The latter stages won’t be done until the feature “sticks” on the Stable channel for a milestone (or so) in the enabled-by-default state (i.e. this won’t start until M141 or M142 at the earliest).  In other words, if sharing this news elsewhere, please be careful in how it is phrased and avoid overpromising how broadly Rust PNG will get deployed in the short-term.

In M136 Rust PNG is disabled by default, but can be enabled through a field trial.  The data below is based on enabling Rust PNG for 1% of Chrome Stable release channel, and comparing against a control group which uses the legacy, `libpng`-based implementation.

# Decoding runtime

If enabled through a field trial, Rust PNG is used to decode PNG images used to render web pages, but also to decode other PNG images (favicons, UI elements, etc).  The `Renderer4.ImageDecodeTaskDurationUs.Png.Gpu` metric captures the runtime of decoding images needed to render web pages.  Changes of this metric look as follows:

* Desktop platforms: improvement
    - Improvement at 75th (around -5%), 95th (around -16%), and 99th percentile (around -16%)
    - Regression at 25th (around +9%) and 50th percentile (around +4%).
    - Overall improvement: sum of all datapoints improved by around -15% (i.e. regression at lower percentiles was overcompensated by improvements at higher percentiles)
    - It seems that the regression at 25th (and maybe also 50th) percentile could be addressed by supporting differently sized `fdeflate` tables.  This wasn’t done because the impact doesn’t seem to support additional costs (increased binary size + extra code/maintenance complexity).  See also https://crbug.com/399441466.
* Android: regression
    - Regression at all percentiles: 25th (around +2%), 50th (around +3%), 75th around (+8%), 95th (around +9%), 99th (around +9%)
    - Overall regression: sum regressed +8.99%
    - This regression is not well understood at this point.
        - One hypothesis would be that the LLVM / `rustc` pipeline results in faster code for x64 than for ARM64, but the desktop data doesn’t support this hypothesis - desktop data restricted to ARM64 doesn’t show a regression (although the improvements are a bit smaller than for desktop overall - only around -5% at 95th percentile and around -9% at 99th percentile).
        - Another unconfirmed hypothesis is that Android is somehow more sensitive to the requirement to page-in Rust stdlib when cold.  One supporting argument for this hypothesis is that a similarly Android-specific performance trouble was observed when shipping Rust-based JSON parser.

# Encoding performance

If enabled through a field trial, Rust PNG is used to encode PNG images in various scenarios: `HTMLCanvasElement.toDataURL`, blobs, tab snapshots, etc.  The metrics below cover only the `HTMLCanvasElement.toDataURL`-related scenarios, which use the following encoding parameters:
* `SkPngEncoder::Options::fZLibLevel`: `3` (see [here](https://source.chromium.org/chromium/chromium/src/+/main:third_party/blink/renderer/platform/image-encoders/image_encoder.cc;l=63;drc=70343319512a4883095fdb100796ea70fba34b9e);  note that `fFilterFlags` is ignored by [`ConvertToRustOptions`](https://source.chromium.org/chromium/chromium/src/+/main:skia/rusty_png_feature.cc;l=21-38;drc=70343319512a4883095fdb100796ea70fba34b9e), but in the end both the control and the experiment groups use non-adaptive filtering)
* `SkPngRustEncoder::CompressionLevel::kLow`
* [`rust_png::Compression::Fastest`](https://docs.rs/png/0.18.0-rc/png/enum.Compression.html#variant.Fastest)

Image encoding speed: mixed results (based on the `Blink.Canvas.ToDataURLScaledDuration.PNG` metric):
* Desktop platforms: regression
    - Regression at 25th (around +24%), 50th (around +14%), 75th (around +6%) percentiles.
    - Not enough data to evaluate the changes at 95th and 99th percentile
* Android: improvement
    - Improvement at 25th (around -9%), 50th (around -7%), 75th (around -6%) percentiles.
    - Not enough data to evaluate the changes at 95th and 99th percentile

Image compression ratio: mixed results (based on the `Blink.Canvas.ToDataURLCompressionRatio.PNG` metric):
* Desktop platforms: statistical issues make experiment-vs-control comparison difficult (samples count in the experiment group is around +4% higher than in the control group)
* Android: regression at 25th (around +20%), 50th (around +12%), 75th (around +11%), 95th (around +2%) percentiles

**Reply by Shnatsel** [reply]
Thanks for the data, that's really helpful!

Do I understand correctly that the encoding numbers are aggregated across all compression levels in use? In the future it would be *very* helpful to disentangle those two modes and measure the performance and compression ratio for each one individually.

[rust_png::Compression::Fastest](https://docs.rs/png/0.18.0-rc/png/enum.Compression.html#variant.Fastest) should absolutely smoke anything libpng has to offer, so seeing a speed regression there would be very surprising. On the other hand, zlib compression level 3 is expected to be kinda slow because it's currently backed by `miniz_oxide`, which outperforms stock zlib but cannot hold a candle to zlib-ng.

There are four ways we could go about improving encoding performance for zlib level 3:

1. According to [fintelia's measurements](https://github.com/image-rs/fdeflate/pull/29#issue-2540760901), Rust PNG's DEFLATE compression level 3 is **not** equivalent to zlib's compression level 3. The closest equivalent to zlib-3 is miniz_oxide-2, which takes only a very small hit in compression ratio on PNG data but is 50% faster than miniz_oxide-3 currently in use in Chromium's integration. Simply translating zlib-3 to miniz_oxide-2 is easiest and is going to be a big performance win.
1. Enable the `zlib-rs` feature on the `png` crate which will use https://github.com/memorysafety/zlib-rs for encoding (not decoding). This is easiest but comes at a cost of some handwritten `unsafe` code, so it doesn't provide as rigorous memory safety guarantees as the rest of `png` crate.
1. Work on optimizing `miniz_oxide`; there should be some decently low-hanging fruit there, since it's a port of miniz C implementation that puts simplicity above performance, but it exposes a C API so there may be limits on how far it can be pushed without breaking ABI compatibility.
1. @fintelia has [prototyped](https://github.com/image-rs/fdeflate/pull/29) a higher compression ratio mode for `fdeflate`. It wasn't really pursued because there was no compelling use case for it. Performance is roughly equivalent to zlib-3 and miniz_oxide-2, but offers a better compression ratio. This is closest to zlib-rs compression level 4 or 5, so if Chromium's fork of zlib is equally optimized and doesn't touch compression ratios then this still may favor compression ratio too much in its current form.

Since encoding always involves a trade-off between speed and compression ratio, it would be helpful to know what our optimization targets are for each compression mode (fastest and zlib-3 respectively).

**Reply by Shnatsel** [reply]
Wait, I've just realized that `SkPngEncoder::Options::fZLibLevel` set to `3` does not translate to [`rust_png::DeflateCompression::level`](https://docs.rs/png/0.18.0-rc/png/enum.DeflateCompression.html) set to 3. There are many intermediate mappings that fudge these parameters, including based on preprocessor macros:

https://source.chromium.org/chromium/chromium/src/+/main:skia/rusty_png_feature.cc;l=21-31;drc=70343319512a4883095fdb100796ea70fba34b9e

https://source.chromium.org/chromium/chromium/src/+/main:third_party/skia/experimental/rust_png/encoder/impl/SkPngRustEncoderImpl.cpp;l=47-63;drc=8a0420284b944fc9e6073137a86e83a24e045e3e

Could you clarify what compression modes in terms of `png` crate parameters are actually used in these field trials?

**Reply by fintelia** [reply]
Thanks for all this data! 

The encoding performance also stood out to me, and I think I figured out what's going on: Chromium is using the Rust streaming encoder, which [always uses flate2 compression](https://github.com/image-rs/image-png/blob/33afddab77449bcd93b1783d2d0ca8ba744cc3c3/src/encoder.rs#L1402) even if you request fdeflate. _Edit:_ created #586

**Reply by 197g** [reply]
Regarding the Desktop's percentile split, is there data available about which specific processor models are on which side of the improvement/regression? It would be nice to find a patter (for instance avx2 is improved and sse3 is not). Rust makes it quite nice to dispatch the proper target features at runtime and have multiple alternatives and the baseline instruction set might not turn out to be an insurmountable hurdle.

**Reply by Shnatsel** [reply]
Regarding the Android-specific regression: the Android target is not very well tested by upstream Rust. It is not unheard of for Android to be missing optimizations in the Rust standard library, see e.g. https://gendignoux.com/blog/2022/11/09/rust-simd-detect-arm-android.html

Given that it's isolated to Android and not ARM, and that Rust-based JSON was also affected, it very well might be a missed `cfg` somewhere in the Rust stdlib or a missing option in the Android target definition in the compiler causing worse performance across the board.

Another hypothesis is that 32-bit ARM still exists on Android but not on desktop (or does it?), so a regression for 32-bit ARM would show up as Android-specific. IIRC you've already explicitly enabled NEON for 32-bit builds? Maybe we have something else that's slow on 32-bit ARM, we have never optimized for 32-bit platforms.

The 32-bit hypothesis could be tested by compiling the crate for a 32bit+NEON target and for a 64bit target and comparing their performance on the same 64-bit hardware. Sadly I don't have any ARM machines so I cannot test it myself.

To check for Android-specific issues with the compiler, I think it's best to profile the decoder on Android and on desktop and compare the flame graphs to see if anything stands out as taking far more time on Android than on desktop. This could point us to inefficiencies in some stdlib functions or at specific functions in png code that we could look into and compare the disassembly between platforms. The tricky bit is to capture the profile with enough samples, so that the discrepancies are not just random noise, as Android devices usually have slow storage and perf profiles can write a lot of data very quickly.

**Reply by fintelia** [reply]
It would make sense if 32-bit systems were the problem. The fdeflate inner decoding loop relies on doing [a 64-bit unaligned load](https://github.com/image-rs/fdeflate/blob/df5729def9dbab605eb6d691b19e90241bbd47b3/src/decompress.rs#L1039) and then extracts bits from the loaded value via shifts and masks. And [comparing the assembly on armv7 vs aarch64](https://godbolt.org/z/oosaW7qEa), it does look like compiling for 32-bit does take way more instructions.

**Reply by awxkee** [reply]
> Regarding the Android-specific regression: the Android target is not very well tested by upstream Rust. It is not unheard of for Android to be missing optimizations in the Rust standard library, see e.g. https://gendignoux.com/blog/2022/11/09/rust-simd-detect-arm-android.html

This looks outdated. On my side, everything currently works with `std::arch::is_aarch64_feature_detected!` without reinventing the wheel. I can say the same for iOS — everything was fixed sometime around mid-2024.

Most of android chips just are `32-bit compatible`, but run in the 64-bit mode. I'm not sure that market share of real 32-bit devices is at least 0.1%. Likely it should be less.

The issue is likely in budget chips like [Arm Cortex-A53](https://en.wikipedia.org/wiki/ARM_Cortex-A53) ( note it can run in 2 modes 32 and 64, but in fact 32 is unlikely used ) that is 12 years old. And those chips also have a long record of performance bottlenecks in places where you don't expect it nowadays. Just from my perspective Rust auto-vectorizer doesn't target them even they are the most used ARM chips in the android devices at the moment.

**Reply by Shnatsel** [reply]
@fintelia that makes me wonder: can we replace the 64-bit decoding loop with a 128-bit one? Will that be beneficial? 128-bit loads are available in the baseline x86_64 and Aarch64. I believe Chromium also requires NEON on 32-bit ARM, so they're available there too.

**Reply by fintelia** [reply]
That would be an interesting idea to check. We'd be able to remove the couple `bit_buffer.fill_buffer()` calls in the middle of the decoding loop because the worst case 3x 12-bit literals + 48-bit length/dist pair = 84-bits would fit in 128-bits.

I'd expect a lot to depend on whether there was extra overhead moving from SIMD registers to integer registers to use the bits in the buffer. Another thing to be aware of is that the decoding loop usually consumes far less than 64-bits. The reason we refill unconditionally for each loop iteration is because a hard-to-predict branch on whether the buffer needs to be refilled (which fires every 5-6 iterations) is more expensive than simply doing the small amount of arithmetic and the memory load.

**Reply by Shnatsel** [reply]
> It would be nice to find a patter (for instance avx2 is improved and sse3 is not). Rust makes it quite nice to dispatch the proper target features at runtime and have multiple alternatives and the baseline instruction set might not turn out to be an insurmountable hurdle.

I've already looked into multiversioning. The bottom line is that it doesn't really help. The filters that do benefit are already more than fast enough, while Paeth does not benefit from SIMD wider than 128 bits. So multiversioning filters is a bust (at least in the decoding step). 

`fdeflate` does benefit from `-C target-cpu=native`, at least on Zen 4, but the gains seem quite diffuse. There is no single function we can multiversion to improve performance. We would need to duplicate most or all of `fdeflate`, and it would be very awkward to do without something like [struct target features](https://github.com/rust-lang/rfcs/blob/01c44d38f5e22ea79a09aa91e067955891a61dea/text/3525-struct-target-feature.md) which are nightly-only if implemented at all.

However, there **is** a way to push performance a little further with SIMD. We have seen some gains from explicit SIMD in decoding for Paeth on x86 with SSE2 compared to the autovectorized implementation. Now that SIMD intrinsics are usable without `unsafe` since Rust v1.87, we can resurrect the formerly portable-SIMD Paeth implementation and rewrite it in terms of x86 intrinsics. SSE2 is included in baseline x86_64 and Rust's i686 target also guarantees SSE2 on recent compiler versions, so we don't even need multiversioning.

**Reply by awxkee** [reply]
> `fdeflate` does benefit from `-C target-cpu=native`, at least on Zen 4, but the gains seem quite diffuse. There is no single function we can multiversion to improve performance. We would need to duplicate most or all of `fdeflate`, and it would be very awkward to do without something like [struct target features](https://github.com/rust-lang/rfcs/blob/01c44d38f5e22ea79a09aa91e067955891a61dea/text/3525-struct-target-feature.md) which are nightly-only if implemented at all.

Since `target_feature` activates featurse in inlined path the solutions below work. It also a little awkward, but not that awkward as duplicating whole crate. Sometimes LLVM rejects to inline it for unknown reasons ( p.s. ```inline(always)``` is just a hint, it's actually not always ) when method contains generics, if that happens then solution `2` usually helps. Inlined methods may be reasonably deep, but when and why LLVM will stop inlining them, that is not really predictable. 

<details>
<summary>Solution 1</summary>

```rust
#[target_feature(enable = "avx2")]
unsafe fn avx_example_parent_method() {
    some_implementation()
}

#[target_feature(enable = "avx2", enable = "fma")]
unsafe fn avx_example_parent_method_with_fma() {
    some_implementation()
}

#[inline(always)]
fn some_implementation() {
         /// Actual work is going here   
}
```

</details>

<details>
<summary>Solution 2</summary>

```rust
#[target_feature(enable = "avx2")]
unsafe fn avx_example_parent_method() {
    let unit = ExecUnit::<1> {

    };
    unit.execute();
}

#[target_feature(enable = "avx2", enable = "fma")]
unsafe fn avx_example_parent_method_with_fma() {
    let unit = ExecUnit::<1> {
        
    };
    unit.execute();
}

struct ExecUnit<const N: usize> {
    
}

impl<const N: usize> ExecUnit<N> {
    #[inline(always)]
    fn execute(&self) {
     /// Actual work is going here   
    }
} 
```

</details>

**Reply by anforowicz** [reply]
> The [Android] issue is likely in budget [ARM] chips

> Regarding the Desktop's percentile split, is there data available about which specific processor models are on which side of the improvement/regression?

I am not able to filter the results to specific processor models.  I *am* able to filter out by the OS version, and the Android decoding regression is almost twice as big on (presumably older) devices running Android 12 or 13, than on devices running Android 15 or 16.  I am also able to filter out by RAM, and Android decoding regression is much more pronounced on devices with <=2GB of RAM than on devices with >=8GB of RAM.

I assume that the desktop regression at 25th and 50th percentile is associated with performance characteristics of `fdeflate` when handling small inputs (see https://crbug.com/399441466).  I assume that the regression at lower percentiles are less correlated with the CPU processor model or characteristics.  (FWIW the sampled decoding times in the 25th percentile are measured in microseconds, but the ones in the 99th percentile are measured in milliseconds.)

> IIRC you've already explicitly enabled NEON for 32-bit builds

Yes - this happened in https://crrev.com/c/6354419 which should be [part of M136](https://chromiumdash.appspot.com/commit/077621684778ae2db179eb040d91bfb389598d38).

> Another hypothesis is that 32-bit ARM still exists on Android but not on desktop

Correct - no 32-bit ARM on desktop and still small, but sizable population of 32-bit ARM on Android.

FWIW the regression is much more pronounced on 32-bit ARM Android (more than 20% across all percentiles).  On 64-bit ARM Android the decoding regression is smaller (less than 3-4% across all percentiles).

**Reply by fintelia** [reply]
Could the problem on ARM be the L1 data cache size? `fdeflate` uses ~18 KB of decoding tables, so performance is going to be far worse if the CPU only has 16 KB of cache and they don't all fit.

**Reply by awxkee** [reply]
@fintelia There are definitely some devices with a 16KB L1 cache — [link to spec](https://www.slideshare.net/OsamaKhalid30/one-plus-5t-architecture). As for whether the problem lies with those phones, I'm not the right person to say. However, I believe such a small cache is uncommon.

**Reply by Shnatsel** [reply]
@anforowicz FYI, a regression in Rust 1.87 affects negatively affects performance of adler32 on x86 CPUs with AVX-2: https://github.com/rust-lang/rust/issues/142603
It will be fixed in v1.88

There are also other performance issues that came with LLVM version upgrade: https://github.com/image-rs/image-png/issues/598

**Reply by Shnatsel** [reply]
@anforowicz I've noticed that [`simd-adler32`](https://github.com/mcountryman/simd-adler32) crate does not have a NEON impl; there's a `neon.rs` file but it's a placeholder with a copy of the AVX-512 implementation. Adding one would be an easy win for ARM performance.

**Reply by anforowicz** [reply]
> @anforowicz I've noticed that [`simd-adler32`](https://github.com/mcountryman/simd-adler32) crate does not have a NEON impl; there's a `neon.rs` file but it's a placeholder with a copy of the AVX-512 implementation. Adding one would be an easy win for ARM performance.

Thanks for flagging this.  @Sentimentron, is this something that you may be interested in?

**Reply by Sentimentron** [reply]
Apologies for missing your message. I can take a look at that

**Reply by Sentimentron** [reply]
Tried a few variants, scalar Adler32 tops out at about ~3 GiB/s on Apple Silicon. Porting the Chromium zlib implementation over seems to raise that to about 33 GiB/s, and I suspect it could go even faster with some of the newer instructions.

**Reply by Sentimentron** [reply]
PR open at https://github.com/mcountryman/simd-adler32/pull/26

**Comment by Shnatsel** [comment]
https://security.googleblog.com/2025/11/rust-in-android-move-fast-fix-things.html makes it sound like the switch has already happened, citing https://chromium-review.googlesource.com/c/chromium/src/+/6085801 from June.

@anforowicz since it's your CL, could you clarify its status?  Has that reached the stable channel yet?

**Reply by anforowicz** [reply]
Yes, sorry for the radio silence - Rust PNG is used in Chromium since M139 (which started shipping to the Chrome Stable release channel in August).  I am now working on also switching PDFium to use Rust PNG and removing `libpng` from Chromium (at least from the product code, some test dependencies are likely to remain).
